src/components/RichText.ts

```diff
@@ -0,0 +1,178 @@
+                }),
+                style: this.props.style
+            },
+            DOM.div({ className: "widget-rich-text-quill", ref: this.setQuillNode })
+        );
+    }
+
+    componentDidMount() {
+        this.renderEditor(this.props);
+    }
+
+    componentWillReceiveProps(nextProps: RichTextProps) {
+        this.updateEditor(nextProps);
+    }
+
+    componentWillUnmount() {
+        this.handleSelectionChange();
+    }
+
+    private setQuillNode(node: HTMLElement) {
+        this.quillNode = node;
+    }
+
+    private renderEditor(props: RichTextProps) {
+        if (this.quillNode && !this.quill) {
+            this.quill = new Quill(this.quillNode, {
+                modules: this.getEditorModules(),
+                theme: props.theme
+            });
+
+            this.quill.on("selection-change", this.handleSelectionChange);
+
+        }
+    }
+
+    private handleSelectionChange() {
+        if (this.quill && !this.quill.hasFocus() && this.props.onChange) {
+            this.props.onChange((this.quill as any).container.firstChild.innerHTML);
+        }
+    }
+
+    private updateEditor(props: RichTextProps) {
+        if (this.quill) {
+            this.quill.enable(!props.readOnly);
+            this.quill.clipboard.dangerouslyPasteHTML(props.value);
+        }
+    }
+
+    private getEditorModules(): Quill.StringMap {
+        if (this.props.editorMode === "basic") {
+            return RichText.getBasicOptions();
+        }
+        if (this.props.editorMode === "advanced") {
+            return RichText.getAdvancedOptions();
+        }
+
+        return RichText.getCustomOptions(this.props.customOptions ? this.props.customOptions : null);
+    }
+
+    private static getBasicOptions(): Quill.StringMap {
+        return {
+            toolbar: [
+                [ "bold", "italic", "underline" ],
+                [ { list: "ordered" }, { list: "bullet" } ]
+            ]
+        };
+    }
+
+    private static getAdvancedOptions(): Quill.StringMap {
+        return {
+            toolbar: [
+                [ RichText.quillOptions.headings ],
+
+                [ "bold", "italic", "underline", "strike" ],
+                [ "blockquote", "code-block" ],
+
+                [ { list: "ordered" }, { list: "bullet" } ],
+                [ { script: "sub" }, { script: "super" } ],
+                [ { indent: "-1" }, { indent: "+1" } ],
+                [ { direction: "rtl" } ],
+
+                [ { color: [] }, { background: [] } ],
+                [ { align: [] } ],
+
+                [ "clean" ]
+            ]
+        };
+    }
+
+    private static getCustomOptions(options: Array<{ option: string }> | null) {
+        const toolbar: { toolbar?: any[] } = {};
+        if (options && options.length) {
+            toolbar.toolbar = [ ...this.processCustomOptions(options) ];
+        }
+
+        return toolbar;
+    }
+
+    private static processCustomOptions(options: Array<{ option: string }>): any[] {
```

Comment:
Removed it, cos it was not very useful since we only ever check. The other time we use an index signature to retrieve values.